classes/helper.php

```diff
@@ -113,4 +116,56 @@ public static function path_validate(string $path) {
 
         return get_string('path_invalid', 'tool_dataflows', $path, true);
     }
+
+    /**
+     * Gets the list of permitted directories that steps are allowed to interact with. The [dataroot]
+     * placeholder will be substituted with the correct dir.
+     *
+     * @return array
+     */
+    public static function get_permitted_dirs(): array {
+        global $CFG;
+
+        $permitteddirs = explode(PHP_EOL, trim(get_config('tool_dataflows', 'permitted_dirs')));
+        foreach ($permitteddirs as &$dir) {
+            if (substr($dir, 0, 10) == '[dataroot]') {
+                $dir = $CFG->dataroot . substr($dir, 10);
+            }
+        }
+        return $permitteddirs;
+    }
+
+    /**
+     * Determines if the given path is a valid filepath.
+     * Derived from https://stackoverflow.com/a/12126772.
+     *
+     * @param $path string The path to check.
```

Comment:
```suggestion
     * @param string $path The path to check.
```
The type should come before the variable name